### ISSUE TITLE
Add support for private key operations based on hardware security devices

### ISSUE THREAD

**Issue by hko-s** [rpgp/rpgp#299]
Currently, there is no mechanism to conveniently use hardware cryptographic devices with rpgp. Adding one would be nice.

Potential use cases include:
- [OpenPGP card](https://gnupg.org/ftp/specs/OpenPGP-smart-card-application-3.4.1.pdf) devices (I've written some [client library code](https://gitlab.com/openpgp-card/openpgp-card) for those)
- PKCS#11 devices
- TPMs

@wiktor-k has signaled interest in co-thinking about this

**Comment by wiktor-k** [comment]
I think this is closely related to https://github.com/rpgp/rpgp/issues/286.

If rpgp could accept signers and decryptor implementing objects (using standard RustCrypto interfaces) we could *potentially* plug other projects such as [PKCS#11](https://github.com/parallaxsecond/rust-cryptoki/pull/192).

One tricky bit is that for PGP signing a couple of things from the PGP "public key" (certificate) are needed, e.g. signing key ID, for embedded that in subpackets (which is not available for any generic keys and couldn't be computed from the raw public key only).

For decryption, up until now, the certificate of the recipient was also needed due to the flexibility in ECDH design (which allowed to put KDF params in the certificate). This has been [fixed in v6](https://gitlab.com/openpgp-wg/rfc4880bis/-/merge_requests/183) (the ticket was from times when it was called v5) but I guess there are still keys in the wild that use non-optimal/standard combinations.

Phew, this is kind of random babbling from my side... I hope that it didn't confuse anyone (too much) :sweat_smile:

**Comment by dignifiedquire** [comment]
I think there might be an opportunity to refactor the internal `crypto` code to be abstract over the rust crypto traits, given all backing impls are using it effectively

**Comment by hko-s** [comment]
> One tricky bit is that for PGP signing a couple of things from the PGP "public key" (certificate) are needed, e.g. signing key ID, for embedded that in subpackets (which is not available for any generic keys and couldn't be computed from the raw public key only).

It's definitely inconvenient that calculating the hash digest for an OpenPGP signature requires additional data from the public key packet (that is not available on all HSM devices).

However, as I understand it, that's an orthogonal issue to providing a mechanism that allows performing raw signing operation on hardware devices?

So I think we have two somewhat separate design tasks:
- Low level signature and decryption operations (e.g. signing a raw hash digest)
- Mechanisms in rpgp that combine these low level operations with the surrounding OpenPGP operation (e.g. calculating that hash digest, which in the case on some HSM can require both the HSM itself, and a copy public key packet data)

**Comment by wiktor-k** [comment]
> I think there might be an opportunity to refactor the internal `crypto` code to be abstract over the rust crypto traits, given all backing impls are using it effectively

That'd be really nice. On the lowest-level there is no difference if one plugs the RSA keypair from RustCrypto or any other implementing the same traits.

Btw, cool to see so much work being done on rpgp! I think I'm not the only one seeing that: https://chaos.social/@kubikpixel/111998260496315248

**Comment by hko-s** [comment]
For the low level cryptographic operations, I see one (multifaceted) concern that need to be addressed somehow:

Authorization on the device. In some cases, a password needs to be provided to the device from the host computer, in other cases, a card reader my require pin entry on an external pin pad, or a device may require touch confirmation.

It's currently unclear to me if these requirements have any implications for the design we're discussing here. However, doing this "right" would be nice (the missing notification facility for touch confirmation in particular strikes me as an unnecessary papercut in existing OpenPGP card UX)

**Comment by wiktor-k** [comment]
Hi folks :wave: 

I've spent a bit of time and got the card signer to work with rpgp. Lo and behold: https://github.com/wiktor-k/test-rpgp/blob/main/src/main.rs and... it works with my key!

(If you want to try it out note that it signs the `DATA` constant and uses my KeyID in `SubpacketData::Issuer` and passes the PIN as a command line argument. Oh, did I mention it supports only ed25519? :sweat_smile: ).

A couple of notes:
  - it'd be cool if the `SecretKeyTrait` was a bit leaner. Just check out the amount of functions I had to mock (via [`#[unimpl]`](https://crates.io/crates/unimpl)) as the code only invokes `create_signature` (or a separate trait is introduced that has only that function),
  - I didn't use `key_pw` but maybe that'd be a good way to read the PIN from the user,
  - I didn't use `hash` but if it was an RSA key it could be necessary.

And that's it!

Btw, I started writing the signing code based [on this example](https://docs.rs/pgp/latest/pgp/composed/signed_key/index.html) but only after a while noticed the `The following is NOT compliant` warning :sweat_smile: I think it'd be better if the example worked (eg by using SignatureConfig as I did) not to proliferate broken code... or maybe it's just me not paying attention :roll_eyes: 

I could contribute some stuff if what I'm saying is making sense :)

**Comment by hko-s** [comment]
Based on the changes done a few months back, hardware security devices are now usable with rPGP: For example, there is now https://codeberg.org/openpgp-card/rpgp, which offers [openpgp-card](https://crates.io/crates/openpgp-card) support for use with rPGP.

**Comment by wiktor-k** [comment]
I think it's safe to close this ticket. Unless some other actionable goal is defined (e.g. "documentation on how to use hardware security device").

**Comment by hko-s** [comment]
As the issue starter, I think closing is appropriate. My goal was to be able to implement https://crates.io/crates/openpgp-card-rpgp, and that's now done.

If we feel the urge to do more related work, we can always open a new issue :)
